src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs

```diff
@@ -54,15 +54,32 @@ public AnalyzerFileReference(string fullPath, IAnalyzerAssemblyLoader assemblyLo
         {
             CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
 
-            FullPath = fullPath;
             _assemblyLoader = assemblyLoader ?? throw new ArgumentNullException(nameof(assemblyLoader));
+            FullPath = getAssemblyLocation(fullPath, assemblyLoader);
 
             _diagnosticAnalyzers = new(this, typeof(DiagnosticAnalyzerAttribute), GetDiagnosticsAnalyzerSupportedLanguages, allowNetFramework: true);
             _generators = new(this, typeof(GeneratorAttribute), GetGeneratorSupportedLanguages, allowNetFramework: false, coerceFunction: CoerceGeneratorType);
 
             // Note this analyzer full path as a dependency location, so that the analyzer loader
             // can correctly load analyzer dependencies.
             assemblyLoader.AddDependencyLocation(fullPath);
+
+            static string getAssemblyLocation(string fullPath, IAnalyzerAssemblyLoader assemblyLoader)
+            {
+                // We use reflection because AnalyzerAssemblyLoader can come from a different assembly (it's source-shared).
```

Comment:
What is the case where `AnalyzerFileReference` doesn't have access to `IAnalyzerAssemblyLoader` members that are `internal`?

Reply:
`AnalyzerFileReference` doesn't have access to `AnalyzerAssemblyLoader.RedirectAssemblyPathExternally` because there are two `AnalyzerAssemblyLoader`s:
- one in Microsoft.CodeAnalysis, that one is fine and AnalyzerFileReference can access it,
- one in Microsoft.CodeAnalysis.Workspaces (it's a different type, only the source code is shared), this is the one actually passed here, but AnalyzerFileReference cannot access it